File: PandorasBox/Ahorn/entities/laseremitter.jl
module PandorasBoxLaserEmitter

using ..Ahorn, Maple

@mapdef Entity "pandorasBox/laserEmitter" LaserEmitter(x::Integer, y::Integer, flag::String="", color::String="White", direction::String="Right", inverted::Bool=false, beamDuration::Int=-1)

const directions = String["Left", "Right", "Up"]
const colors = sort(collect(keys(Ahorn.XNAColors.colors)))

const placements = Ahorn.PlacementDict(
    "Laser Emitter ($direction) (Pandora's Box)" => Ahorn.EntityPlacement(
        LaserEmitter,
        "point",
        Dict{String, Any}(
            "direction" => direction
        )
    ) for direction in directions
)

Ahorn.editingOptions(entity::LaserEmitter) = Dict{String, Any}(
    "direction" => directions,
    "color" => colors
)

sprite = "objects/pandorasBox/laser/emitter/idle0"

function Ahorn.selection(entity::LaserEmitter)
    x, y = Ahorn.position(entity)

    return Ahorn.getSpriteRectangle(sprite, x, y, jx=0.5, jy=1.0)
end

Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::LaserEmitter, room::Maple.Room) = Ahorn.drawSprite(ctx, sprite, 0, 0, jx=0.5, jy=1.0)

end

File: PandorasBox/Loenn/entities/laser_emitter.lua
-- TODO - Dropdowns

local laserEmitter = {}

local directions = {
    "Left",
    "Right",
    "Up"
}

laserEmitter.name = "pandorasBox/laserEmitter"
laserEmitter.depth = 50
laserEmitter.fieldInformation = {
    color = {
        fieldType = "color",
        allowXNAColors = true,
    }
}
laserEmitter.placements = {}

for _, direction in ipairs(directions) do
    table.insert(laserEmitter.placements, {
        name = string.lower(direction),
        data = {
            flag = "",
            color = "White",
            direction = direction,
            inverted = false,
            beamDuration = -1
        }
    })
end

laserEmitter.texture = "objects/pandorasBox/laser/emitter/idle0"
laserEmitter.justification = {0.5, 1.0}

return laserEmitter